roc.go

```diff
@@ -43,35 +44,42 @@ func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []floa
 	if !sort.Float64sAreSorted(y) {
 		panic("stat: input must be sorted")
 	}
-
-	var incWidth, tol float64
-	if n == 0 {
-		if len(y) == 0 {
-			return nil, nil
-		}
-		tpr = make([]float64, len(y)+1)
-		fpr = make([]float64, len(y)+1)
-	} else {
-		if n < 2 {
-			panic("stat: cannot calculate fewer than 2 points on a ROC curve")
-		}
-		if len(y) == 0 {
-			return nil, nil
-		}
-		tpr = make([]float64, n)
-		fpr = make([]float64, n)
-		incWidth = (y[len(y)-1] - y[0]) / float64(n-1)
-		tol = y[0] + incWidth
-		if incWidth == 0 {
-			tpr[n-1] = 1
-			fpr[n-1] = 1
-			return
+	if !sort.Float64sAreSorted(cutoffs) {
+		panic("stat: input must be sorted")
+	}
+	if len(y) == 0 {
+		return nil, nil
+	}
+	var bin int
+	if len(cutoffs) == 0 {
+		cutoffs = make([]float64, len(y)+1)
+		cutoffs[0] = math.Nextafter(y[0], y[0]-1)
+		// Choose all possible cutoffs but remove duplicate values
+		// in y.
+		for i, u := range y {
+			if i != 0 && u != y[i-1] {
+				bin++
+			}
+			cutoffs[bin+1] = u
 		}
+		cutoffs = cutoffs[0:(bin + 2)]
 	}
 
-	var bin int = 1 // the initial bin is known to have 0 fpr and 0 tpr
+	tpr = make([]float64, len(cutoffs))
+	fpr = make([]float64, len(cutoffs))
+	bin = 0
 	var nPos, nNeg float64
 	for i, u := range classes {
+		// Update the bin until it matches the next y value
+		// (skip empty bins).
+		for (bin < len(cutoffs)) && (y[i] > cutoffs[bin]) {
```

Comment:
parens, cool.

if we wanted to force the ROC curve to always start at (0,0) and always end at (1,1) then your suggestion would work -- but this implementation allows for arbitrary cutoffs to be defined and so that is no longer necessarily the case. You could even pass a single cutoff to get a single fpr, tpr pair. Now I'm thinking about it there are alot of extra special cases I should add tests for...